ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -51,54 +66,119 @@ def loop(self):
         while not rospy.is_shutdown():
             if self.pose and self.waypoints and self.waypoints_2d and self.waypoint_tree:
                 # Get closest waypoint
-                closest_waypoint_idx = self.get_closest_waypoint_idx()
-                self.publish_waypoints(closest_waypoint_idx)
-            rate.sleep()
-
-    def get_closest_waypoint_idx(self):
-        x = self.pose.pose.position.x
-        y = self.pose.pose.position.y
-        closest_idx = self.waypoint_tree.query([x, y], 1)[1]
-
-        # Check if closest is ahead or behind vehicle
-        closest_coord = self.waypoints_2d[closest_idx]
-        prev_coord = self.waypoints_2d[closest_idx - 1]
-
-        # Equation for hyperplane through closest_coords
-        cl_vect = np.array(closest_coord)
-        prev_vect = np.array(prev_coord)
-        pos_vect = np.array([x, y])
+                self.closest_wp_idx = self.get_closest_waypoint_idx()
+                self.drive_state_machine()
+                self.publish_waypoints()
 
-        val = np.dot(cl_vect - prev_vect, pos_vect - cl_vect)
+            rate.sleep()
 
-        if val > 0:
-            # closest_coord is behind us
-            closest_idx = (closest_idx + 1) % len(self.waypoints_2d)
-        return closest_idx
+    def drive_state_machine(self):
```

Comment:
Can you document this?

Reply:
Regarding speed wobbling and crashing, I have observed this too occasionally.. I'm pretty sure it is due to frame rate .. though I could not verify with 50Hz either on my PC